File: api/app/landcover.py
"""Bucket FortyGuard satellite class percents into planner language."""

from __future__ import annotations

from typing import Any


def bucket_classes(percents: dict[str, float] | None) -> dict[str, Any]:
    percents = percents or {}
    canopy = _sum_matching(percents, ("tree", "plant", "canopy"))
    # plant is already in canopy; vegetation adds grass/shrub without double-counting trees
    grass = _sum_matching(percents, ("grass", "shrub", "vegetation"))
    impervious = _sum_matching(percents, ("building", "road", "route", "sidewalk", "pavement", "asphalt"))
    water = _sum_matching(percents, ("water",))
    return {
        "canopy_pct": round(canopy, 2),
        "vegetation_pct": round(canopy + grass, 2),
        "impervious_pct": round(impervious, 2),
        "water_pct": round(water, 2),
        "classes_percent": percents,
        "note": (
            "Satellite mix at the hotspot centroid. tree/plant ≈ canopy; "
            "building/road/sidewalk ≈ impervious. Not a district-wide NLCD raster."
        ),
    }


def _sum_matching(percents: dict[str, float], tokens: tuple[str, ...]) -> float:
    total = 0.0
    for name, value in percents.items():
        lower = name.lower()
        if any(tok in lower for tok in tokens):
            try:
                total += float(value)
            except (TypeError, ValueError):
                continue
    return total

File: web/src/lib/landcover.ts
/** Hotspot satellite mix for the sidebar stacked bar. Not a district NLCD raster. */

export type MixSlice = { key: string; label: string; pct: number; color: string };

const COLORS = {
  canopy: "#a3e635",
  vegetation: "#2dd4bf",
  impervious: "#94a3b8",
  water: "#22d3ee",
  other: "#3f4a5a",
};

type Buckets = {
  canopy_pct: number;
  vegetation_pct: number;
  impervious_pct: number;
  water_pct: number;
};

function sumMatching(percents: Record<string, number>, tokens: readonly string[]): number {
  let total = 0;
  for (const [name, value] of Object.entries(percents)) {
    const lower = name.toLowerCase();
    if (!tokens.some((tok) => lower.includes(tok))) continue;
    const n = Number(value);
    if (Number.isFinite(n)) total += n;
  }
  return total;
}

export function satelliteMixSlices(
  buckets?: Buckets | null,
  classes?: Record<string, number> | null,
): MixSlice[] {
  let canopy = 0;
  let vegetation = 0;
  let impervious = 0;
  let water = 0;
  if (buckets) {
    canopy = buckets.canopy_pct || 0;
    // vegetation_pct is canopy + grass; stack must not double-count trees.
    vegetation = Math.max(0, (buckets.vegetation_pct || 0) - canopy);
    impervious = buckets.impervious_pct || 0;
    water = buckets.water_pct || 0;
  } else if (classes && Object.keys(classes).length) {
    canopy = sumMatching(classes, ["tree", "plant", "canopy"]);
    vegetation = sumMatching(classes, ["grass", "shrub"]);
    impervious = sumMatching(classes, ["building", "road", "route", "sidewalk", "pavement", "asphalt"]);
    water = sumMatching(classes, ["water"]);
  } else {
    return [];
  }
  const slices: MixSlice[] = [
    { key: "canopy", label: "Canopy", pct: canopy, color: COLORS.canopy },
    { key: "vegetation", label: "Vegetation", pct: vegetation, color: COLORS.vegetation },
    { key: "impervious", label: "Impervious", pct: impervious, color: COLORS.impervious },
    { key: "water", label: "Water", pct: water, color: COLORS.water },
  ].filter((s) => s.pct > 0.05);
  if (!slices.length) return [];
  const used = slices.reduce((sum, s) => sum + s.pct, 0);
  if (used < 99.5) {
    slices.push({ key: "other", label: "Other", pct: Math.max(0, 100 - used), color: COLORS.other });
  }
  return slices;
}
